br/pkg/lightning/backend/local/local.go

```diff
@@ -1630,43 +1293,76 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID, regi
 
-	for {
-		unfinishedRanges := lf.unfinishedRanges(ranges)
-		if len(unfinishedRanges) == 0 {
-			break
+	var (
+		workGroup, workerCtx = errgroup.WithContext(ctx)
+		// jobCh is unbuffered so when we finished sending all jobs, we can make sure all jobs have been
+		// received by workers.
+		jobCh = make(chan *regionJob)
+		// jobWg is the number of jobs that need to be processed by worker in this round.
+		jobWg           sync.WaitGroup
+		needRetryJobs   []*regionJob
+		needRetryJobsMu sync.Mutex
+		putBackFn       = func(job *regionJob) {
+			needRetryJobsMu.Lock()
+			defer needRetryJobsMu.Unlock()
+			needRetryJobs = append(needRetryJobs, job)
 		}
-		log.FromContext(ctx).Info("import engine unfinished ranges", zap.Int("count", len(unfinishedRanges)))
+	)
 
-		// if all the kv can fit in one region, skip split regions. TiDB will split one region for
-		// the table when table is created.
-		needSplit := len(unfinishedRanges) > 1 || lfTotalSize > regionSplitSize || lfLength > regionSplitKeys
-		// split region by given ranges
-		for i := 0; i < maxRetryTimes; i++ {
-			err = local.SplitAndScatterRegionInBatches(ctx, unfinishedRanges, lf.tableInfo, needSplit, regionSplitSize, maxBatchSplitRanges)
-			if err == nil || common.IsContextCanceledError(err) {
-				break
-			}
+	for i := 0; i < local.workerConcurrency; i++ {
+		workGroup.Go(func() error {
+			return local.startWorker(workerCtx, jobCh, &jobWg, putBackFn)
+		})
+	}
 
-			log.FromContext(ctx).Warn("split and scatter failed in retry", zap.Stringer("uuid", engineUUID),
-				log.ShortError(err), zap.Int("retry", i))
-		}
+	var pendingJobs []*regionJob
+	for {
+		pendingJobs = append(pendingJobs, needRetryJobs...)
+		needRetryJobs = nil
+		log.FromContext(ctx).Info("import engine pending jobs", zap.Int("count", len(pendingJobs)))
+
+		newJobs, err := local.prepareAndGenerateUnfinishedJob(
+			ctx,
+			engineUUID,
+			lf,
+			ranges,
+			regionSplitSize,
+			regionSplitKeys,
+		)
 		if err != nil {
-			log.FromContext(ctx).Error("split & scatter ranges failed", zap.Stringer("uuid", engineUUID), log.ShortError(err))
 			return err
 		}
 
-		// start to write to kv and ingest
-		err = local.writeAndIngestByRanges(ctx, lf, unfinishedRanges, regionSplitSize, regionSplitKeys)
-		if err != nil {
-			log.FromContext(ctx).Error("write and ingest engine failed", log.ShortError(err))
-			return err
+		pendingJobs = append(pendingJobs, newJobs...)
+		if len(pendingJobs) == 0 {
+			break
+		}
+
+		jobWg.Add(len(pendingJobs))
+		for _, job := range pendingJobs {
+			select {
+			case <-workerCtx.Done():
+				return workGroup.Wait()
+			case jobCh <- job:
+			}
+		}
+		pendingJobs = nil
+		// all jobs are received by workers and worker will always call jobWg.Done() after processing
+		// no need to check workerCtx.Done() here.
+		jobWg.Wait()
+		// check if worker has error in this round
+		select {
+		case <-workerCtx.Done():
+			return workGroup.Wait()
+		default:
 		}
 	}
 
+	close(jobCh)
```

Comment:
`jobCh` also needs to close before any `return` statement and wait for workGroup to exit.